### PR TITLE
fix: use correct extended attribute to mark directories and files

### LINKED ISSUE #55: Creating nested path hierarchies and empty directories inside the ZIP archive?
Hey, small question.

From what I've observed from the ZIP spec, it seems like the "directory hierarchy" of ZIP files are just an illusion, and that it's really just POSIX-style names (forward slash separator), where each file has its own full path embedded straight into the file "name" data (no "actual nested folders"), such as `name = foo/bar/baz/readme.txt`. I've also noticed that empty directories are stored as "files" which are named things like `foo/bar/baz/` with a trailing slash at the end, and with a DATA LENGTH of ZERO. The trailing slash seems to be the only signifier that a "file" is actually a directory?

For instance, when I created a ZIP with the standard utilities on Linux, I observed that it has an entry for the "parent folder" as a standalone ZIP entry, before it has an entry for the actual file inside that folder.

The hierarchy in this example ZIP is:

- `A folder\with backslashes`: This is the actual folder name, it's a POSIX name so backslash is legal. And it's stored in the ZIP as `A folder\with backslashes/`
- `A folder\with backslashes/my file\withbackslash.txt`: This is the filename within the folder. It's just a file with a backslash in the name.

Here's my test ZIP:

[backslashes.zip](https://github.com/uktrade/stream-zip/files/11840406/backslashes.zip)

So with that backstory about how directories seem to work in ZIPs, I'm very curious about the correct ways to handle directories in `stream-zip`:

1. If I want to create a deeply nested directory, is there anything I need to think about and be careful about? Or can I just immediately do `name = foo/bar/baz/readme.txt` without having to manually create the parent path components (`foo/`, `foo/bar/` and `foo/bar/baz/`)?
2. Is it safe to do interleaved writes, like creating `foo/a/1.txt`, then `bar/b/1.txt`, then `foo/a/2.txt`? Meaning is it safe to add directories in a jumbled order to the ZIP file contents? I assume that `stream-zip` remembers which directories it has already written to the ZIP and doesn't create those paths again.
3. And lastly... is there any way to add empty directories? Sometimes there are empty ones in the data I am packing, and I'd prefer to preserve that empty hierarchy if it's possible. Using some kind of "if directory doesn't exist in ZIP yet, create an empty one in the ZIP"? I thought of possibly creating a file named `the/directory/` with a trailing slash and providing `b""` (empty bytes) as the data?
4. If the idea in point 3 would work, would there be any risk of clashing with or overwriting any actual files/subfolders from that hierachy if they already existed? Or any other failures, such as it generating new directory entries for the same "directory file/" over and over again? (i.e. if `a/b/` was already auto-generated and added earlier, and I then manually try to write `a/b/`, I guess I'd end up with a duplicate directory entry)?
5. If it's not currently possible to add empty directories, is there any chance of some API for adding empty directories? It could be as simple as "if the user provides a tuple which ONLY contains a name and timestamp and nothing else, add that as a directory entry to the ZIP, if it hasn't been added before".

### PR DESCRIPTION
**Pull request by Arcitec**
* Refactors "is directory" into a variable to make the code clearer and to allow easy reuse.

* Now uses the correct bit attribute for marking directories on ZIPs that are marked as "Made on UNIX" (which is what stream-zip uses).

Closes #55.

### CODE REVIEW

**Comment by Arcitec** [comment]
@michalc I did not implement the "throw exception if trying to encode an `is_dir == True` when `chunks` is non-empty". It's vital for the ZIP spec that we never allow any data bytes when encoding a directory.

But I did not find a clean way to implement it, so since you're the expert on the library it's better that I don't attempt to do that! :) ❤️ 

Thanks again for this library, you're absolutely brilliant.

**Comment by michalc** [comment]
One thing I realise about the existing test - it doesn't exactly test any real affect of how the directory-related code works. What does this fix in terms of real-world behaviour? Can there be a test on that?

**Comment by Arcitec** [comment]
> One thing I realise about the existing test - it doesn't exactly test any real affect of how the directory-related code works. What does this fix in terms of real-world behaviour? Can there be a test on that?

Since `stream-zip` marks all ZIPs as "Made by UNIX", this fixes the "is a directory" attribute to what UNIX-created ZIPs should contain.

I am not sure how we could add tests for this since it's a decoder-specific behavior! I'd guess that most unzippers use similar heuristics as 7-Zip, which is that they prefer a trailing slash and only look into the attributes when there isn't a trailing slash. I also assume many unzippers are lazy and don't look at attributes at all!

I'm currently doing some tests with 7-Zip, and the fixed code works:

Before fix:

image

After fix:

image

How I did the test:

```py
from datetime import datetime

from stream_zip import (
    stream_zip,
    ZIP_64,
    ZIP_32,
)

def test_directory_without_slashes_zipfile():
    now = datetime.fromisoformat('2021-01-01 21:01:12')
    perms = 0o600

    def files():
        yield 'file-1', now, perms, ZIP_64, (b"",)
        yield 'file-2', now, perms, ZIP_64, (b"",)
        yield 'file-3', now, perms, ZIP_32, (b"",)
        yield 'file-4', now, perms, ZIP_32, (b"",)

    with open("out.zip", "wb") as f:
        for chunk in stream_zip(files()):
            f.write(chunk)

test_directory_without_slashes_zipfile()
```

ALONG with a "hack" in `stream_zip.py`:

```py
            is_dir = name_encoded[-1:] == b'/'
            is_dir = True
```

Meaning that I made `stream-zip` "treat every name as directory even if no trailing slash".

The issue if we want to automate this test in the test suite is that `stream-zip` only treats things as directory if there's a trailing slash. And if we add a trailing slash, then 7-Zip doesn't even care about the attributes and only looks at the slash to determine "it's a dir".

So the only way I could test if the attribute fix worked was to forcibly encode names WITHOUT trailing slash but WITH the fixed attributes.

And yeah it works beautifully. These are the correct attrs for UNIX dirs. :)

As for the other issue: Should I open a ticket to track/remember that `stream-zip` should throw if `len(bytes) > 0` when `is_dir == True`? It's a low-priority issue but right now the library can make illegal zips very easily by just giving some bytes with a filename that ends in a slash.

**Comment by Arcitec** [comment]
@michalc As an "interesting" sidenote: On Unix, all directories must have the "x" (executable) permission to be readable, even by the owner itself. It's not enough to have "rw" permission, needs to be "rwx". So I considered automatically adding the "x" flag when `is_dir` is true. But in the end, I decided that it's the API user's job to send in correct permissions instead. All we should do as a library is to mark it as a dir. If they want to ZIP a dir that doesn't have a "x" flag then so be it. It's too difficult to guess what people want.

But this does mean that if somebody wants to manually create empty directories, they should give a name ending in `/` AND also give correct permissions (typically `0o755` which is the octal for "rwxr-xr-x" (So everyone can read/execute "visit the dir", and only owner can write to it).

**Comment by Arcitec** [comment]
Just because it's related to all of this directory stuff, I'll just add that I've also verified that it's legal to not add parent directories to the ZIP entries. So going straight for `name = foo/bar/baz/` is legal (there's no need to add `foo/` and `foo/bar/` as separate entries before that). The ONLY benefit of adding separate entries for all directory parents is that it allows specifying the parent folder's modification time and permissions too.

If nothing has been specified, then the parent shows up like this:

image

But that's totally fine. Lots of ZIP utilities skip entries for the parents. 7-Zip adds each parent component. But it's yet another implementation detail which would beup to the user in `stream-zip`'s case. So if the user wants modification times for all folders, they'd simply have to manually keep track of every path they've added to the ZIP so far (a `set()` would be appropriate), and for every directory, add a 0-byte entry with the permissions+modification time of the directory, and the name (ending with `/`).

Like:

```
foo/ (0 bytes, with timestamp and permissions)
foo/bar/ (0 bytes, with timestamp and permissions)
foo/bar/baz (regular file)
```

Just use something like a `set()` to ensure that you *never* add the exact same path (`name`) twice to the archive! Because that would be a conflict.

But for clarity, in case anyone else reads this in the future: You can skip that whole process and just add the files without caring about adding the parents. They'll simply be treated as "folders with unknown metadata" by the unzippers such as 7-Zip.

**Comment by Arcitec** [comment]
@michalc Regarding the other bug, I have an idea for the dir byte sanity check.

Since directories are not allowed to have any data, we could do something like this:

```py
for name, modified_at, perms, method, chunks in files:
    # snip...

    is_dir = name_encoded[-1:] == b'/'
    if is_dir:
        assert next(chunks, None) is None, "directories are not allowed to contain any data bytes"
        chunks = ()
```

https://docs.python.org/3/library/functions.html#next

Basically, check if the iterator is empty and warn them if it isn't. We should throw a proper exception though, and you are the expert on that exception hierarchy and message style. :)

After checking the iterator we should also always force it to be empty, so that there will NEVER be any risk of data written to a directory entry.

What do you think of this idea?

If this solution seems reasonable, it would be a pretty trivial change to do this and add a test to make sure an exception is thrown.

**Comment by Arcitec** [comment]
I had a quick look at this again today and was able to find the official Linux header file where all these magic bit-numbers come from:

https://pubs.opengroup.org/onlinepubs/7908799/xsh/sysstat.h.html

https://github.com/torvalds/linux/blob/a92b7d26c743b9dc06d520f863d624e94978a1d9/include/uapi/linux/stat.h#L14

```cpp
#define S_IFMT  00170000
#define S_IFDIR  0040000
#define S_IFREG  0100000
```

All those numbers are in octal format. And they're standardized for UNIX (plus Linux) since the 1970s. They will never change.

The basic summary is as follows:

- The "mode/perms" bits don't just contain "the 755/644/whatever permission bits". They also contain information about what type of file it is, such as "is directory", "is regular file", "is symbolic link", etc.
- `S_IFMT` is a bitmask for looking at the "file type" bits of the mode. Allows the program to ignore the permission bits.
- `S_IFDIR` is the bit that's set when it's a directory.
- `S_IFREG` is the bit that's set when it's a "regular file".

I also did some tests back on Linux, creating 1 directory (`mkdir foo`) and 1 file (`touch bar`) with default permissions. Then I used Python's `pathlib.Path().stat().st_mode` to extract the actual octal bit values of both kinds.

The results were:

- Directory: `0o40755` (this means `S_IFDIR` with permissions 755 (`rwxr-xr-x`)). So a folder that's readable and visitable (executable) by anyone. Writable by the owner.
- File: `0o100644` (this means `S_IFREG` with permissions 644 (`rw-r--r--`)). So a file that's readable by anyone. Writable by the owner.

So this pull request takes care of marking the `S_IFDIR` bit correctly when adding directories to the ZIP.

It does not take care of marking `S_IFREG` for "regular file" correctly when adding files to the ZIP. But I have an idea for a solution for that and will take care of that now.

**Comment by Arcitec** [comment]
@michalc Alright, everythign here is ready for final review now. :)

After discovering the official names for the magic numbers, and finding that Python had a library for those constants, it was just a matter of refactoring. I've now also added support for fixing the file type bits for both directories **and** files. And added unit tests.

If there's anything else you'd like to change, please let me know.

**Comment by michalc** [comment]
(I've been without much access to a computer this past couple of weeks, but should be able to look into this now)

> I am not sure how we could add tests for this since it's a decoder-specific behavior!

Ah - I would use the specific decoders in the test. The existing tests use Python's zipfile module, infozip's unzip, stream-unzip, libarchive, and 7-zip. If this addresses issues with directories not being created properly when a zip is unzipped with 7-zip, then I think the best test would be one that invokes 7-zip and asserts that the file/directory structure it creates is right.

I know lots of the existing tests assert on external_attr from Python's zipfile module, but I don't really think that's what we care about here. Directories being created by 7-zip I think is what we care about, and so if possible, should be what the test asserts. I see the existing test_directory_zipfile test - but I think we should be able to do better, especially if bugs have been revealed. It's a bit... testing on an implementation detail rather than what I think is cared about - whether directories are created when unzipped.

File: stream_zip.py [hunk 1]
```diff
@@ -358,7 +359,7 @@ def _no_compression_32_local_header_and_data(name_encoded, mod_at_encoded, exter
 
         def _no_compression_buffered_data_size_crc_32(chunks, maximum_size):
             # We cannot have a data descriptor, and so have to be able to determine the total
-            # length and CRC32 before output ofchunks to client code
+            # length and CRC32 before output of chunks to client code
```
Comment (michalc) [inline stream_zip.py:362]: I'm being slightly nit-picky, but can this be in another commit (or even, another PR that we can merge in quickly)? To make the history just that tiny bit clearer.

**Comment by michalc** [comment]
On the sentinel idea for the directory iterable, I think I'm a bit anti really, because:

We can still raise an exception if it's not empty if we want without this
It make types a bit more complex. I know there isn't any type annotation, but right now the iterable is an iterable of bytes. It would be an iterable of bytes or the sentinal.
In any case, I'm good with any exceptions not being in this PR (or any supporting code or TODOs that mention them)

**Comment by michalc** [comment]
Also, I'm wondering if this is good enough:

```python
is_dir = name_encoded[-1:] == b'/'
st_mode_with_type = st_mode | (stat.S_IFDIR if is_dir else stat.S_IFREG)
external_attr = \
    (st_mode_with_type << 16) | \
    (0x10 if is_dir else 0x0)  # MS-DOS directory
```

I know it doesn't check if the type is specified by client code, but is that... fine? What ill effects would this cause?

**Comment by michalc** [comment]
I've started to wonder: should just be up to client code, and do the bare minimum of magic in the library mode-wise, i.e. no setting of file type at all.

So a regular file would have its mode passed in by client code as:
```
stat.S_IFREG | 0o600
```
and a directory should end in a forward slash and have its mode passed in as:
```
stat.S_IFDIR | 0o700
```
If the file is a directory according to the mode, stream-zip could do what's necessary to make it a directory zip-wise for various clients. Specifically, it can add a trailing slash if it doesn't have one, and set the MS-DOS directory flag.

Then:

- It's consistent with how symbolic links go in - the user has to specify it in the mode passed in
- It's how directories are identified/specified in the *nix world, so typical knowledge can be applied
- The mode specified by the client code == mode in the ZIP in all cases, so it's not surprising on that front

Although it does have the downside of changing the name of empty directories, by adding a trailing slash. So maybe a bit surprising on that front. But given it sounds like this is what other ZIP writers do anyway, suspect that's probably fine. It has the bonus of keeping the requirements down API-wise to make a directory.

It's also technically a breaking change - the MS-DOS directory flag is added under a different condition. But I guess I would be a bit surprised if this effects even one user at this point - someone out there would be making zip files with empty directories that are to be read by ancient clients?

The code would then be:

```python
is_dir = is_dir = stat.S_ISDIR(st_mode)
name_with_trailing_slash_if_needed = os.path.join(name, '') if is_dir else name
name_encoded = name_with_trailing_slash_if_needed.encode('utf-8')
external_attr = \
    (st_mode << 16) | \
    (0x10 if is_dir else 0x0)  # MS-DOS directory
```
which I don't think is too bad in terms of complexity

**Comment by michalc** [comment]
To communicate, I'm a bit tempted to document the behaviour as it is now to make directories "correctly": pass in the right mode, and end the path with a forward slash. Nothing proposed in this PR or above I think would make this wrong

**Comment by michalc** [comment]
> To communicate, I'm a bit tempted to document the behaviour as it is now to make directories "correctly": pass in the right mode, and end the path with a forward slash. Nothing proposed in this PR or above I think would make this wrong

Although I'm already stuck on what to call these things in the documentation. We have 2 "modes" - one is the low level linux/file mode, and one is the "zip" mode...

**Comment by michalc** [comment]
Maybe this is how the 5 properties should be described:

1. File name
2. Modification time
3. Mode (File type and permissions)
4. Type (Compression and metadata format)
5. Binary contents

**Comment by michalc** [comment]
Although it mentions "type" twice...

**Comment by michalc** [comment]
Ah the spec uses "method", so maybe this:

1. File name
2. Modification time
3. Mode (File type and permissions)
4. Method (Compression mechanism)
5. Binary contents

**Comment by michalc** [comment]
(Repeating the comment I left in https://github.com/uktrade/stream-zip/issues/55)

I've opted to go for a documentation approach to this in https://github.com/uktrade/stream-zip/pull/60, documenting the directories must end in a forward slash, and should have the S_IFDIR mode. Potentially a code change could be made down the line, but I'm tempted to leave it for now since from what I can tell, this works in all clients, and is The Right Thing in terms of what ends up in the file.
